File: Assignment-2/2/knapsack.py
"""
    ***************SAFE MOVE****************
    => Sort value per unit in decending order
    ***************ALGORITHM****************
    => Insert the largest values[i]/weights[i] to the
        knapsack if it is feasible else fill it as much
        as it can hold
    => Insert the next as per the above rule and continue
"""
def min_val(num_one, num_two):
    """
        Returns minimum of two values
    """
    if num_one <= num_two:
        return num_one
    return num_two

def get_max_index(value_per_unit, no_of_items):
    """
        Gets the maximum index of the item having high value per unit
    """
    max_index = 0
    for i in range(0, no_of_items - 1):
        if value_per_unit[i] >= value_per_unit[max_index]:
            max_index = i
    return max_index

def get_max_value(no_of_items, total_weight, values, weights):
    """
        Actual logic goes here
    """
    value_per_unit = list()
    for i in range(0 , no_of_items):
        value_per_unit.append(values[i] / weights[i])
    #taken_amount = 0
    final_value = 0
    for i in range(0 , no_of_items):
        if total_weight == 0:
            return final_value
    
        max_index = get_max_index(value_per_unit, no_of_items)
        taken_amount = min_val(weights[max_index] , total_weight)
        final_value = final_value + (taken_amount * value_per_unit[max_index])
        weights[max_index] = weights[max_index] - taken_amount
        if weights[max_index] == 0:
            value_per_unit[max_index] = 0
        else:
            value_per_unit[max_index] = values[max_index] - weights[max_index]
        total_weight = total_weight - taken_amount
    return final_value


def start():
    """
        Program starts here
    """
    #no_of_items = int(input("Give total number of items"))
    no_of_items = int(input())
    #total_weight = int(input("Give total capacity of knapsack"))
    total_weight = int(input())

    weights = list()
    values = list()
    #print("Give values and weights")
    temp_value = 0
    temp_weight = 0
    for i in range(0, no_of_items):
        temp_value = int(input())
        temp_weight = int(input())
        weights.append(temp_weight)
        values.append(temp_value)
    #print("Maximum value holded by knapsack:")
    print("%.4f" %get_max_value(no_of_items, total_weight, values, weights))

if __name__ == "__main__":
    start()

File: Assignment-2/2/Knapsack.java
import java.util.Scanner;
import java.text.DecimalFormat;

/*
    ***************SAFE MOVE****************
    => Sort value per unit in decending order
    ***************ALGORITHM****************
    => Insert the largest values[i]/weights[i] to the
        knapsack if it is feasible else fill it as much
        as it can hold
    => Insert the next as per the above rule and continue
*/

public class Knapsack
{
    private static int min(int numOne, int numTwo)
    {
        if(numOne <= numTwo)
            return numOne;
        return numTwo;
    }

    private static int getMaxIndex(double valuePerUnit[], int noOfItems)
    {
        int maxIndex = 0;
        for(int i = 0; i < noOfItems; i++)
        {
            if(valuePerUnit[maxIndex] <= valuePerUnit[i])
                maxIndex = i;
        }
        return maxIndex;
    }

    private static double getMaxValue(int noOfItems, int totalWeight, int weights[], int values[])
    {
        double valuePerUnit[] = new double[noOfItems];
        for(int i = 0; i < noOfItems; i++)
        {
            valuePerUnit[i] = ((double)values[i]) / weights[i]; 
        }
        double finalValue = 0;
        for(int i = 0; i < noOfItems; i++)
        {
            if(totalWeight == 0)
                return finalValue;
            int maxIndex = getMaxIndex(valuePerUnit, noOfItems);
            int takenAmount = min(weights[maxIndex] , totalWeight);
            finalValue += takenAmount * valuePerUnit[maxIndex];
            weights[maxIndex] -= takenAmount;
            if(weights[maxIndex] == 0)
                valuePerUnit[maxIndex] = 0;
            else
                valuePerUnit[maxIndex] = values[maxIndex] / weights[maxIndex];
            totalWeight -= takenAmount;
        }
        return finalValue;
    }

    public static void main(String args[])
    {
        Scanner scanner = new Scanner(System.in);
        DecimalFormat decimalFormat = new DecimalFormat("#.####");
        
        //System.out.println("Give total number of items:");
        int noOfItems = scanner.nextInt();

        //System.out.println("Give total capacity of knapsack:");
        int totalWeight = scanner.nextInt();

        int values[] = new int[noOfItems];
        int weights[] = new int[noOfItems];

        //System.out.println("Give values and their weights:");
        for(int i = 0; i < noOfItems; i++)
        {
            values[i] = scanner.nextInt();
            weights[i] = scanner.nextInt();
        }

        //System.out.println("Maximum value can be stored in knapsack is:");
        System.out.println(decimalFormat.format(getMaxValue(noOfItems, totalWeight, weights, values)));
        scanner.close();
    }
}